yapapi/runner/__init__.py

```diff
@@ -310,22 +311,25 @@ async def find_offers():
                     raise
                 async for proposal in events:
                     emit_progress(ProposalEvent.RECEIVED, proposal.id, from_=proposal.issuer)
-                    score = await strategy.score_offer(proposal)
-                    if score < SCORE_NEUTRAL:
-                        proposal_id, provider_id = proposal.id, proposal.issuer
-                        with contextlib.suppress(Exception):
-                            await proposal.reject()
-                        emit_progress(ProposalEvent.REJECTED, proposal_id, for_=provider_id)
-                        continue
-                    if proposal.is_draft:
-                        emit_progress(ProposalEvent.BUFFERED, proposal.id)
-                        offer_buffer[proposal.issuer] = _BufferItem(datetime.now(), score, proposal)
-                    else:
-                        try:
+                    try:
+                        score = await strategy.score_offer(proposal)
```

Comment:
So this change is to also catch exceptions from `strategy.score_offer(proposal)`? Are they due to a malformed proposal? If so, then I'd surround just this line with `try ... catch` (preferably with some more specific exception type) and reject the proposal (by setting low score`).

Reply:
The `score_offer` has been addressed by @filipgolem in 903cfc0 (I agree with more specific exception, prolly some kind of `JsonParsingError`).

But I would still want to have a general catch all handler here to avoid unnecessary hanging of an example due to malformed proposal or just some troll-node-provider. In first iteration I logged the actual exception but was advised against it by @filipgolem .

Reply:
Good point. Actually I now like the idea of not catching exceptions in `score_offer` better, so that we can catch them here and emit `ProposalEvent.REJECTED` with `reason=str(exc)`, where `exc` is the caught exception.

Reply:
As I'm looking at the "event emitter":
https://github.com/golemfactory/yapapi/blob/f1fcfd372a0d78f9106f6068c50ccd5fa5bc4f66/yapapi/runner/events.py#L158-L160
I see that msg is constructed manuallish, and you can't just provide `exc_info=True` (like in stdlib logger) to dump a traceback. IMHO just the name of an exception (i.e. `str(e)`) is not enough.

Reply:
`str(e)` is the exception's message, not the name, but you're right that reporting unexpected exceptions (the ones that are caused by bugs in `yapapi`) by emitting events is not a good idea, `Engine.map()` should throw them to the external code.

Reply:
Summing up, here's what I would do (I changed my mind on this few times before arriving at the conclusion):

Since `score_offer()` is user code that can do anything, let's not catch **any** exceptions from `strategy.score_offer(proposal)` but let's make sure they're re-thrown from `Engine.map()` (that's the tricky part) so that the user can deal with them.

Alternatively, we may create a new exception `InvalidProposalError` and catch only this exception when calling `strategy.score_offer(proposal)`:
```
try:
    score = await strategy.score_offer(proposal)
except InvalidProposalError as exc:
    emit_progress(ProposalEvent.REJECTED, proposal_id, reason=exc)
```

Reply:
noooice. I think it's good direction

Reply:
Exceptions from `score_offer` shouldn't stop the `find_offers` coroutine from running, so they have to be caught in `find_offers`. Your alternative solution (catching InvalidProposalError from `score_offer`) is a good idea.

Reply:
> Exceptions from `score_offer` shouldn't stop the `find_offers` coroutine from running

I'm not sure if I agree. We're talking about unexpected errors that result from bugs in `score_offer` or some low-level code that it uses.